File: Non-procedural Programming/Solved/antihric/male/grupa.hs
import Functional

--http://forum.matfyz.info/viewtopic.php?f=169&t=11043&p=40642&hilit=grupa#p40642 (tady to je v prologu, ale bylo to min 1x uz zadano i na haskell)

testGrupa1 = [
    ['e', 'a', 'b'], --1
    ['a', 'b', 'e'], --3
    ['b', 'e', 'a']  --3 
    ]

testGrupa2 = [
    ['e', 'a', 'b', 'c'], --1
    ['a', 'b', 'e', 'c'], --4
    ['b', 'c', 'a', 'e'], --3
    ['c', 'e', 'c', 'b']  --3
    ]

testGrupa3 = [
    ['e', 'a', 'b', 'c'], --1
    ['a', 'b', 'e', 'c'], --4
    ['b', 'c', 'a', 'b'], --3
    ['c', 'e', 'c', 'e']  --2 
    ]

transform :: [[Char]] -> ([Char], [[Char]])
transform vstup = (h,transpose c) 
    where
        h = head vstup
        a = tail vstup
        b = transpose a
        c = tail b

op :: ([Char],[[Char]]) -> Char -> Char -> Char
op (h,v) a b = case i of
                Just ii ->  case j of
                            Just jj -> (v !! (ii-1)) !! (jj-1)
                            Nothing -> error "Neplatne pismenko."

                Nothing -> error "Neplatne pismenko."
        where
            i = findIndex (\x -> x==a) h
            j = findIndex (\x -> x==b) h

rad :: ([Char],[[Char]]) -> Char -> Int
rad _ 'e' = 1
rad v a = rad' v 1 a a
            where
                rad' :: ([Char],[[Char]]) -> Int -> Char -> Char -> Int
                rad' _ n 'e' _ = n
                rad' v n a b = rad' v (n+1) c b
                    where
                        c = op v a b

File: Non-procedural Programming/Solved/antidvorak/male/grupa.pl
?- consult('../functional.pl').

vstup(V) :-
    V = [
        [e, a, b], %1
        [a, b, e], %3
        [b, e, a]  %3 
        ].

vstup2(V) :-
    V = [
       [e, a, b, c], %1
       [a, b, e, c], %4
       [b, c, a, e], %3
       [c, e, c, b]  %3
       ].

vstup3(V) :-
   V = [
      [e, a, b, c], %1
      [a, b, e, c], %4
      [b, c, a, b], %3
      [c, e, c, e]  %2 
      ].

%http://forum.matfyz.info/viewtopic.php?f=169&t=11043&p=40642&hilit=grupa#p40642

transform(V, H-Z) :-
    V = [H | Y],
    transpose(Y,YT),
    YT = [_ | TT],
    transpose(TT, Z).

op(_, A, e, A).
op(_, e, A, A).
op(H-Z, A, B, C) :-
    findIndex(unify(A), H, I),
    findIndex(unify(B), H, J),
    nth1(I, Z, R),
    nth1(J, R, C).

rad(_, e, 1) :- !.
rad(V, A, R) :- transform(V, H-Z), rad(H-Z, 1, A, A, R).

rad(_, N, e, _, N).
rad(H-Z, N, A, B, R) :-
    op(H-Z, A, B, C),
    M is N+1,
    rad(H-Z, M, C, B, R), !.